dartagnan/src/main/java/com/dat3m/dartagnan/parsers/cat/VisitorBase.java

```diff
@@ -0,0 +1,347 @@
+        Relation r1 = relation(c.e);
+        return addDefinition(new TransitiveClosure(r0, r1));
+    }
+
+    @Override
+    public Relation visitExprTransRef(ExprTransRefContext c) {
+        checkNoCurrent(c);
+        Relation r0 = newRelation(alias());
+        Relation r1 = wmm.newRelation();
+        Relation r2 = wmm.getRelation(ID);
+        Relation r3 = relation(c.e);
+        return addDefinition(new TransitiveClosure(r0, addDefinition(new Union(r1, r2, r3))));
+    }
+
+    @Override
+    public Relation visitExprDomainIdentity(ExprDomainIdentityContext c) {
+        checkNoCurrent(c);
+        Relation r0 = newRelation(alias());
+        Relation r1 = relation(c.e);
+        return addDefinition(new DomainIdentity(r0, r1));
+    }
+
+    @Override
+    public Relation visitExprRangeIdentity(ExprRangeIdentityContext c) {
+        checkNoCurrent(c);
+        Relation r0 = newRelation(alias());
+        Relation r1 = relation(c.e);
+        return addDefinition(new RangeIdentity(r0, r1));
+    }
+
+    @Override
+    public Relation visitExprOptional(ExprOptionalContext c) {
+        checkNoCurrent(c);
+        Relation r0 = newRelation(alias());
+        Relation r1 = relation(c.e);
+        return addDefinition(new Union(r0, wmm.getRelation(ID), r1));
+    }
+
+    @Override
+    public Relation visitExprIdentity(ExprIdentityContext c) {
+        checkNoCurrent(c);
+        Relation r0 = newRelation(alias());
+        FilterAbstract s1 = set(c.e);
+        return addDefinition(new Identity(r0, s1));
+    }
+
+    @Override
+    public Relation visitExprCartesian(ExprCartesianContext c) {
+        checkNoCurrent(c);
+        Relation r0 = newRelation(alias());
+        FilterAbstract s1 = set(c.e1);
+        FilterAbstract s2 = set(c.e2);
+        return addDefinition(new CartesianProduct(r0, s1, s2));
+    }
+
+    @Override
+    public Relation visitExprFencerel(ExprFencerelContext ctx) {
+        checkNoCurrent(ctx);
+        Relation r0 = newRelation(alias());
+        FilterAbstract s1 = set(ctx.e);
+        return addDefinition(new Fences(r0, s1));
+    }
+
+    private void addCurrentName(Object o) {
+        if (alias == null) {
+            return;
+        }
+        if (o instanceof Relation) {
+            wmm.addAlias(alias, (Relation) o);
+        }
+        namespace.put(alias, o);
+        alias = null;
+    }
+
+    private void checkNoCurrent(ExpressionContext c) {
+        if(current != null) {
+            throw new ParsingException("Unexpected recursive context at expression: " + c.getText());
+        }
+    }
+
+    private Relation current() {
+        Relation r = current;
+        current = null;
+        return r;
+    }
+
+    private void checkNoCurrentOrNoAlias(ExpressionContext c) {
+        verify(current == null || alias == null,
+                "Broken structure invariant at expression: {}", new Object() {@Override public String toString() { return c.getText(); }});
```

Comment:
Why does this code generate an instance of an anonymous class?

Reply:
To avoid calling `getText()` when the assertion holds.  I also misused log4j's formatting syntax here.

Reply:
If you want to avoid calling `getText()`, I would rather do a normal if-conditional to check for the invariant (just like `checkNoCurrent` does).